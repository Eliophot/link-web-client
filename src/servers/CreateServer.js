import React, { useEffect } from 'react';
import { v4 as uuid } from 'uuid';
import PropTypes from 'prop-types';
import NoMenuLayout from '../common/NoMenuLayout';
import { ServerForm } from './helpers/ServerForm';
import './CreateServer.scss';

const SHOW_IMPORT_MSG_TIME = 4000;
const propTypes = {
  createServer: PropTypes.func,
  history: PropTypes.shape({
    push: PropTypes.func,
  }),
  resetSelectedServer: PropTypes.func,
};

const CreateServer = (ImportServersBtn, useStateFlagTimeout) => {
  const CreateServerComp = ({ createServer, history: { push }, resetSelectedServer }) => {
    const [ serversImported, setServersImported ] = useStateFlagTimeout(false, SHOW_IMPORT_MSG_TIME);
    const handleSubmit = (serverData) => {
      const id = uuid();
      const server = { id, ...serverData };

      createServer(server);
      push(`/server/${id}/list-short-urls/1`);
    };

    useEffect(() => {
      resetSelectedServer();
    }, []);

    return (
      <NoMenuLayout>
        <ServerForm onSubmit={handleSubmit}>
          <ImportServersBtn onImport={setServersImported} />
          <button className="btn btn-outline-primary">Create server</button>
        </ServerForm>

        {serversImported && (
          <div className="row create-server__import-success-msg">
            <div className="col-md-10 offset-md-1">
              <div className="p-2 mt-3 bg-main text-white text-center">
                Servers properly imported. You can now select one from the list :)
              </div>
            </div>
          </div>
        )}
      </NoMenuLayout>
    );
  };

  CreateServerComp.propTypes = propTypes;

  return CreateServerComp;
};

export default CreateServer;
