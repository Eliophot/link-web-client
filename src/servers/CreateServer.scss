@import '../utils/base';

.create-server__label {
  font-weight: 700;
  cursor: pointer;

  @media (min-width: $mdMin) {
    text-align: right;
  }
}

.create-server__csv-select {
  position: absolute;
  left: -9999px;
  top: -9999px;
}
