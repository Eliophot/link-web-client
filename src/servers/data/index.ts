export interface ServerData {
  name: string;
  url: string;
  apiKey: string;
}

export interface ServerWithId extends ServerData {
  id: string;
}

export interface ReachableServer extends ServerWithId {
  version: string;
  printableVersion: string;
}

export interface NonReachableServer extends ServerWithId {
  serverNotReachable: true;
}

export interface NotFoundServer {
  serverNotFound: true;
}

export type RegularServer = ReachableServer | NonReachableServer;

export type SelectedServer = RegularServer | NotFoundServer | null;

export type ServersMap = Record<string, ServerWithId>;

export const hasServerData = (server: ServerData | NotFoundServer | null): server is ServerData =>
  !!(server as ServerData)?.url && !!(server as ServerData)?.apiKey;

export const isReachableServer = (server: SelectedServer): server is ReachableServer =>
  !!server?.hasOwnProperty('printableVersion');

export const isServerWithId = (server: SelectedServer | ServerWithId): server is ServerWithId =>
  !!server?.hasOwnProperty('id');
