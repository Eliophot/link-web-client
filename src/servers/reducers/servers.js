import { handleActions } from 'redux-actions';
import { pipe, assoc, map, reduce, dissoc } from 'ramda';
import { v4 as uuid } from 'uuid';

/* eslint-disable padding-line-between-statements */
export const EDIT_SERVER = 'shlink/servers/EDIT_SERVER';
export const DELETE_SERVER = 'shlink/servers/DELETE_SERVER';
export const CREATE_SERVERS = 'shlink/servers/CREATE_SERVERS';
/* eslint-enable padding-line-between-statements */

const initialState = {};

const assocId = (server) => assoc('id', server.id || uuid(), server);

export default handleActions({
  [CREATE_SERVERS]: (state, { newServers }) => ({ ...state, ...newServers }),
  [DELETE_SERVER]: (state, { serverId }) => dissoc(serverId, state),
  [EDIT_SERVER]: (state, { serverId, serverData }) => !state[serverId]
    ? state
    : assoc(serverId, { ...state[serverId], ...serverData }, state),
}, initialState);

export const createServer = (server) => createServers([ server ]);

const serversListToMap = reduce((acc, server) => assoc(server.id, server, acc), {});

export const createServers = pipe(
  map(assocId),
  serversListToMap,
  (newServers) => ({ type: CREATE_SERVERS, newServers })
);

export const editServer = (serverId, serverData) => ({ type: EDIT_SERVER, serverId, serverData });

export const deleteServer = ({ id }) => ({ type: DELETE_SERVER, serverId: id });
