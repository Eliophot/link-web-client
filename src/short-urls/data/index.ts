import * as m from 'moment';
import { Nullable, OptionalString } from '../../utils/utils';

export interface EditShortUrlData {
  longUrl?: string;
  tags?: string[];
  title?: string;
  validSince?: m.Moment | string | null;
  validUntil?: m.Moment | string | null;
  maxVisits?: number | null;
  validateUrl?: boolean;
}

export interface ShortUrlData extends EditShortUrlData {
  longUrl: string;
  customSlug?: string;
  shortCodeLength?: number;
  domain?: string;
  findIfExists?: boolean;
}

export interface ShortUrl {
  shortCode: string;
  shortUrl: string;
  longUrl: string;
  dateCreated: string;
  visitsCount: number;
  meta: Required<Nullable<ShortUrlMeta>>;
  tags: string[];
  domain: string | null;
  title?: string | null;
}

export interface ShortUrlMeta {
  validSince?: string;
  validUntil?: string;
  maxVisits?: number;
}

export interface ShortUrlModalProps {
  shortUrl: ShortUrl;
  isOpen: boolean;
  toggle: () => void;
}

export interface ShortUrlIdentifier {
  shortCode: string;
  domain: OptionalString;
}
