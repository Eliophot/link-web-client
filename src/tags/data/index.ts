import type { ShlinkVisitsSummary } from '../../api/types';

export interface TagStats {
  shortUrlsCount: number;
  visitsCount: number;
  visitsSummary?: ShlinkVisitsSummary; // Optional only before Shlink 3.5.0
}

export interface TagModalProps {
  tag: string;
  isOpen: boolean;
  toggle: () => void;
}

export interface SimplifiedTag {
  tag: string;
  shortUrls: number;
  visits: number;
}
