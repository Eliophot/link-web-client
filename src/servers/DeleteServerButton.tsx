import React, { FC } from 'react';
import { faMinusCircle as deleteIcon } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useToggle } from '../utils/helpers/hooks';
import { DeleteServerModalProps } from './DeleteServerModal';
import { ServerWithId } from './data';

export interface DeleteServerButtonProps {
  server: ServerWithId;
  className?: string;
  textClassName?: string;
}

const DeleteServerButton = (DeleteServerModal: FC<DeleteServerModalProps>): FC<DeleteServerButtonProps> => (
  { server, className, children, textClassName },
) => {
  const [ isModalOpen, , showModal, hideModal ] = useToggle();

  return (
    <React.Fragment>
      <span className={className} onClick={showModal}>
        {!children && <FontAwesomeIcon icon={deleteIcon} />}
        <span className={textClassName}>{children ?? 'Remove this server'}</span>
      </span>

      <DeleteServerModal server={server} isOpen={isModalOpen} toggle={hideModal} />
    </React.Fragment>
  );
};

export default DeleteServerButton;
