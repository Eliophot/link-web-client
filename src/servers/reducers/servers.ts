import { handleActions } from 'redux-actions';
import { pipe, assoc, map, reduce, dissoc } from 'ramda';
import { v4 as uuid } from 'uuid';
import { NewServerData, ServerWithId } from '../data';

/* eslint-disable padding-line-between-statements */
export const EDIT_SERVER = 'shlink/servers/EDIT_SERVER';
export const DELETE_SERVER = 'shlink/servers/DELETE_SERVER';
export const CREATE_SERVERS = 'shlink/servers/CREATE_SERVERS';
/* eslint-enable padding-line-between-statements */

export type ServersMap = Record<string, ServerWithId>;

const initialState: ServersMap = {};

const serverWithId = (server: ServerWithId | NewServerData): ServerWithId => {
  if ((server as ServerWithId).id) {
    return server as ServerWithId;
  }

  return assoc('id', uuid(), server);
};

export default handleActions<ServersMap, any>({
  [CREATE_SERVERS]: (state, { newServers }: any) => ({ ...state, ...newServers }),
  [DELETE_SERVER]: (state, { serverId }: any) => dissoc(serverId, state),
  [EDIT_SERVER]: (state, { serverId, serverData }: any) => !state[serverId]
    ? state
    : assoc(serverId, { ...state[serverId], ...serverData }, state),
}, initialState);

const serversListToMap = reduce<ServerWithId, ServersMap>((acc, server) => assoc(server.id, server, acc), {});

export const createServers = pipe(
  map(serverWithId),
  serversListToMap,
  (newServers: ServersMap) => ({ type: CREATE_SERVERS, newServers }),
);

export const createServer = (server: ServerWithId) => createServers([ server ]);

export const editServer = (serverId: string, serverData: Partial<NewServerData>) => ({
  type: EDIT_SERVER,
  serverId,
  serverData,
});

export const deleteServer = ({ id }: ServerWithId) => ({ type: DELETE_SERVER, serverId: id });
