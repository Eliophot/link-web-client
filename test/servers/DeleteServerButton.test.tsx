import { ReactNode } from 'react';
import { screen, waitFor } from '@testing-library/react';
import { Mock } from 'ts-mockery';
import { DeleteServerButton as createDeleteServerButton } from '../../src/servers/DeleteServerButton';
import { ServerWithId } from '../../src/servers/data';
import { renderWithEvents } from '../__helpers__/setUpTest';

describe('<DeleteServerButton />', () => {
  const DeleteServerButton = createDeleteServerButton(
    ({ isOpen }) => <>DeleteServerModal {isOpen ? '[Open]' : '[Closed]'}</>,
  );
  const setUp = (children?: ReactNode) => renderWithEvents(
    <DeleteServerButton server={Mock.all<ServerWithId>()} textClassName="button">{children}</DeleteServerButton>,
  );

  it.each([
    ['Foo bar'],
    ['baz'],
    ['something'],
    [undefined],
  ])('renders expected content', (children) => {
    const { container } = setUp(children);
    expect(container.firstChild).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays modal when button is clicked', async () => {
    const { user, container } = setUp();

    expect(screen.getByText(/DeleteServerModal/)).toHaveTextContent(/Closed/);
    expect(screen.getByText(/DeleteServerModal/)).not.toHaveTextContent(/Open/);
    container.firstElementChild && await user.click(container.firstElementChild);

    await waitFor(() => expect(screen.getByText(/DeleteServerModal/)).toHaveTextContent(/Open/));
  });
});
