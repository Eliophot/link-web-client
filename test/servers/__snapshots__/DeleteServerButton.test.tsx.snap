// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeleteServerButton /> renders expected content 1`] = `
<span>
  <span
    class="button"
  >
    Foo bar
  </span>
</span>
`;

exports[`<DeleteServerButton /> renders expected content 2`] = `
<span>
  <span
    class="button"
  >
    baz
  </span>
</span>
`;

exports[`<DeleteServerButton /> renders expected content 3`] = `
<span>
  <span
    class="button"
  >
    something
  </span>
</span>
`;

exports[`<DeleteServerButton /> renders expected content 4`] = `
<span>
  <svg
    aria-hidden="true"
    class="svg-inline--fa fa-circle-minus fa-fw "
    data-icon="circle-minus"
    data-prefix="fas"
    focusable="false"
    role="img"
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 256C0 114.6 114.6 0 256 0C397.4 0 512 114.6 512 256C512 397.4 397.4 512 256 512C114.6 512 0 397.4 0 256zM168 232C154.7 232 144 242.7 144 256C144 269.3 154.7 280 168 280H344C357.3 280 368 269.3 368 256C368 242.7 357.3 232 344 232H168z"
      fill="currentColor"
    />
  </svg>
  <span
    class="button"
  >
    Remove this server
  </span>
</span>
`;
